### PR TITLE
IGNITE-22942 Add ability to StaticNodeFinder to resolve multiple IP addresses under the same host name

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-22942

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/StaticNodeFinder.java [hunk 1]
```diff
@@ -36,8 +52,37 @@ public StaticNodeFinder(List<NetworkAddress> addresses) {
...
+                .collect(toList());
+    }
+
+    private static String[] resolveAll(String host) {
+        InetAddress[] inetAddresses;
+        try {
+            inetAddresses = InetAddress.getAllByName(host);
+        } catch (UnknownHostException e) {
+            LOG.warn("Cannot resolve {}", host);
+            return new String[0];
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/StaticNodeFinder.java:71]: We have a constant for empty arrays)

File: modules/table/src/testFixtures/java/org/apache/ignite/internal/table/TxAbstractTest.java [hunk 2]
```diff
@@ -330,7 +330,12 @@ protected TxManager txManager(TableViewInternal t) {
     protected boolean assertPartitionsSame(TableViewInternal table, int partId) {
         long storageIdx = 0;
 
-        for (Map.Entry<String, Loza> entry : txTestCluster.raftServers().entrySet()) {
+        Map<String, Loza> raftServers = txTestCluster.raftServers();
+        if (raftServers == null) {
```
Comment (sashapolo) [inline modules/table/src/testFixtures/java/org/apache/ignite/internal/table/TxAbstractTest.java:334]: I wonder how you stumbled upon this)

File: modules/table/src/testFixtures/java/org/apache/ignite/distributed/ItTxTestCluster.java [hunk 3]
```diff
@@ -218,7 +218,7 @@ public class ItTxTestCluster {
 
     private ReplicaService clientReplicaSvc;
 
-    protected Map<String, Loza> raftServers;
+    protected @Nullable Map<String, Loza> raftServers;
```
Comment (sashapolo) [inline modules/table/src/testFixtures/java/org/apache/ignite/distributed/ItTxTestCluster.java:221]: Why isn't it initialized to an empty map?

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        NetworkAddress ipv4 = new NetworkAddress("1.2.3.4", 3001);
+        NetworkAddress ipv6 = new NetworkAddress("2001:1234:130f:1234:1234:99c0:876a:130b", 3002);
+        NodeFinder finder = new StaticNodeFinder(List.of(ipv4, ipv6));
+
+        assertThat(finder.findNodes(), contains(ipv4, ipv6));
+    }
+
+    @Test
+    void resolvesLocalHostToJustOneAddress() throws Exception {
+        NodeFinder finder = new StaticNodeFinder(List.of(new NetworkAddress("localhost", 3001)));
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:60]: How do you guarantee that there's more than one address to choose from?

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+import org.apache.ignite.internal.testframework.WorkDirectory;
+import org.apache.ignite.internal.testframework.WorkDirectoryExtension;
+import org.apache.ignite.network.NetworkAddress;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(WorkDirectoryExtension.class)
+class StaticNodeFinderTest {
+    @WorkDirectory
+    private Path workDir;
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:47]: We have a base Ignite test (`AbstractIgniteTest`?) that has a workDir in it

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        return ipToHostname.entrySet().stream()
+                .map(entry -> entry.getKey() + " " + entry.getValue())
+                .collect(joining("\n"));
+    }
+
+    private static String stdoutString(Process process) {
+        try (InputStream stdout = process.getInputStream()) {
+            return new String(stdout.readAllBytes(), UTF_8);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:161]: I think you simply propagate these exceptions

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            Path hostsFilePath
+    ) throws IOException, InterruptedException {
+        String javaBinaryPath = ProcessHandle.current().info().command().orElseThrow();
+        String javaClassPath = System.getProperty("java.class.path");
+
+        //noinspection UseOfProcessBuilder
+        ProcessBuilder processBuilder = new ProcessBuilder(
+                javaBinaryPath,
+                "-cp", javaClassPath,
+                "-Djdk.net.hosts.file=" + hostsFilePath,
```
Comment (sashapolo) [inline modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:131]: `System.setProperty` doesn't work for this?

File: modules/table/src/testFixtures/java/org/apache/ignite/internal/table/TxAbstractTest.java (hunk 2 above)
Comment (rpuch) [reply modules/table/src/testFixtures/java/org/apache/ignite/internal/table/TxAbstractTest.java:334]: "localhost" was being resolved to more than 1 address, so more than one set of managers per node was started, so start was failing due to a busy port. So the start method (which is a `@BeforeEach`) was not able to create the `raftService` map. `@AfterEach` method was run then, which is assuming that the start method finished successfully, so an NPE was thrown masking the original problem.

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java (hunk 5 above)
Comment (rpuch) [reply modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:47]: Is it mandatory to use it? Explicit usage of `@WorkDirectory` is not too cumbersome, and I don't want to subclass without a good reason.

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java (hunk 7 above)
Comment (rpuch) [reply modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:131]: This property has effect first time a name resolver is created per JVM, so it's too late to override it at runtime.

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java (hunk 5 above)
Comment (sashapolo) [reply modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java:47]: nope, that was just a suggestion

File: modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java (hunk 4 above)
Comment (rpuch) [reply modules/network/src/test/java/org/apache/ignite/internal/network/StaticNodeFinderTest.java]: I rewrote it to the approach with overrides to guarantee this

**Review by sashapolo — APPROVED** [review]

**Pull request merged**